src/lsm/grid.zig

```diff
@@ -214,20 +227,67 @@ pub fn GridType(comptime Storage: type) type {
             for (grid.read_iop_tracer_slots) |slot| assert(slot == null);
             for (grid.write_iop_tracer_slots) |slot| assert(slot == null);
 
-            grid.cache.deinit(allocator);
-
             for (grid.cache_blocks) |block| allocator.free(block);
             allocator.free(grid.cache_blocks);
 
+            grid.cache_valid.deinit(allocator);
+            grid.cache.deinit(allocator);
+
             grid.* = undefined;
         }
 
+        pub fn cancel(grid: *Grid, callback: fn (*Grid) void) void {
+            assert(grid.canceling == null);
+
+            grid.canceling = .{ .callback = callback };
+
+            grid.read_pending_queue.reset();
+            grid.superblock.storage.reset_next_tick_lsm();
+            grid.superblock.storage.on_next_tick(
```

Comment:
What are our expectations w.r.t. `storage.next_tick_queue_vsr` at this point?

For example, what if `cancel()` is called successively? Would another concurrent process in `...queue_vsr` compete with our context memory here?

Reply:
We have no expectations of `storage.next_tick_queue_vsr`.

`cancel()` is not called again until the first `cancel()` finishes. i.e. concurrent `cancel`s is not permitted.
```zig
        pub fn cancel(grid: *Grid, callback: fn (*Grid) void) void {
            assert(grid.canceling == null);
```

Reply:
Do we not then expect the `next_tick_queue_vsr` to contain some fixed N (or 1 or 0) items? i.e. That it contains no concurrent cancel callback?

The assert on `grid.canceling` is good but it feels like we should also be able to cross-check this with the next tick queue for `.vsr` (either positively or negatively) as we do elsewhere.

Reply:
> That it contains no concurrent cancel callback?

Right, but we can't actually assert that because the vsr nexttick queue can be used for other nexttick calls which are not correlated in any way to `grid.cancel()`.

Reply:
Gotcha.